### PR TITLE
Add Analytical Custom-JVP Derivative Paths for Soft Robots

### PR DESCRIPTION
**Pull request by mstoelzle**
# PR Description

## Summary

This PR adds analytical derivative paths behind JAX custom JVP rules for `SoftRobot`
systems, migrates the system classes to protected implementation hooks, renames
ambiguous derivative APIs to `*_time_derivative*` and
`*_arc_length_derivative*`, and consolidates derivative benchmarking into
`tools/benchmarks/benchmark_derivative_paths.py`.

The implementation keeps public APIs differentiable with normal JAX calls such as
`jax.grad`, `jax.jvp`, `jax.jacfwd`, and `jax.jacrev`, while allowing the public
methods to dispatch to analytical derivative implementations when custom JVPs are
enabled. A global trace-time toggle (`soromox.autodiff.custom_jvp_mode`) can disable
these rules before JIT tracing.

## Benchmark Setup

Derivative benchmark command:

```bash
uv run python tools/benchmarks/benchmark_derivative_paths.py \
  --systems planar_pcs pcs gvs \
  --segment-counts 1 2 4 8 16 32 \
  --strategies public_custom_jvp_enabled public_custom_jvp_disabled \
  --custom-jvp-modes enabled disabled \
  --execution-repeats 3 \
  --warmup-runs 1 \
  --csv benchmarks/final-derivative-custom-jvp.csv \
  --json benchmarks/final-derivative-custom-jvp.json \
  --markdown-summary benchmarks/final-derivative-custom-jvp.md
```

System-method benchmark command was run on this branch and on `origin/main` with
the matching old/new Jacobian derivative method names. The benchmark now wraps
each measured callable in `jax.jit` before timing. For the `origin/main`
reference, the benchmark was executed from a temporary `git archive` checkout
with the same JIT-wrapping benchmark fix applied to that temporary script.

```bash
uv run python tools/benchmarks/benchmark_system_methods.py \
  --systems planar_pcs pcs gvs \
  --segment-counts 1 2 4 8 16 32 \
  --methods forward_kinematics forward_kinematics_batched jacobian \
    jacobian_bodyframe jacobian_inertialframe jacobian_and_time_derivative \
    jacobian_and_time_derivative_bodyframe \
    jacobian_and_time_derivative_inertialframe inertia_matrix coriolis_matrix \
    gravitational_force forward_dynamics total_energy \
  --execution-repeats 3 \
  --duration 0.02 \
  --solver-dt 0.005 \
  --save-dt 0.01 \
  --csv benchmarks/final-system-methods-current.csv \
  --json benchmarks/final-system-methods-current.json \
  --plot benchmarks/final-system-methods-current.pdf
```

## Custom-JVP Enabled vs Disabled

Speedup is `disabled execution time / enabled execution time`, so values above
`1.00x` mean enabling custom JVPs was faster. `n=32` is the geomean across the
single 32-segment row for each case.

| case | PlanarPCS geomean | PlanarPCS n=32 | PCS geomean | PCS n=32 | GVS geomean | GVS n=32 |
|---|---:|---:|---:|---:|---:|---:|
| `jacobian` | 0.99x | 0.80x | 1.15x | 1.18x | 1.08x | 1.01x |
| `forward_kinematics_jvp_q` | 1.03x | 1.12x | 1.02x | 1.10x | 1.17x | 1.08x |
| `forward_kinematics_jvp_s` | 0.98x | 0.95x | 1.33x | 1.84x | 0.94x | 1.05x |
| `jacobian_jvp_q` | 1.09x | 1.58x | 1.17x | 1.43x | 1.13x | 1.07x |
| `jacobian_jvp_s` | 1.49x | 1.86x | 0.99x | 1.30x | 1.09x | 1.41x |
| `gravitational_energy_gradient` | 0.73x | 0.87x | 0.98x | 0.76x | 1.48x | 1.00x |
| `elastic_energy_gradient` | 1.04x | 0.71x | 1.39x | 0.95x | 0.93x | 0.71x |
| `potential_energy_gradient` | 0.78x | 0.71x | 1.00x | 0.72x | 1.51x | 1.14x |
| `kinetic_energy_gradient` | 1.40x | 2.44x | 1.65x | 2.14x | 1.04x | 1.13x |
| `total_energy_gradient` | 1.32x | 2.33x | 1.39x | 1.16x | 1.09x | 1.10x |

Aggregate geomean speedups across derivative cases and segment counts, excluding
the plain `jacobian` primal wrapper-overhead control:

| system | all cases | n=32 only | strongest geomean win | weakest geomean case |
|---|---:|---:|---|---|
| PlanarPCS | 1.07x | 1.25x | `jacobian_jvp_s` 1.49x | `gravitational_energy_gradient` 0.73x |
| PCS | 1.19x | 1.19x | `kinetic_energy_gradient` 1.65x | `gravitational_energy_gradient` 0.98x |
| GVS | 1.14x | 1.06x | `potential_energy_gradient` 1.51x | `elastic_energy_gradient` 0.93x |

## Primal Custom-JVP Wrapper Overhead

The system-method benchmark below measures regular forward calls under JIT. As
an additional focused check, public calls were measured with custom JVPs enabled
and disabled on this branch. Values are steady-state JIT execution medians.

| system | segments | method | enabled | disabled | overhead |
|---|---:|---|---:|---:|---:|
| PlanarPCS | 1 | `forward_kinematics` | 4.58 us | 4.54 us | +0.9% |
| PlanarPCS | 32 | `forward_kinematics` | 19.79 us | 10.71 us | +84.8% |
| PCS | 1 | `forward_kinematics` | 5.75 us | 5.58 us | +3.0% |
| PCS | 32 | `forward_kinematics` | 49.46 us | 49.17 us | +0.6% |
| GVS | 1 | `forward_kinematics` | 41.63 us | 39.92 us | +4.3% |
| GVS | 32 | `forward_kinematics` | 372.54 us | 379.63 us | -1.9% |
| PlanarPCS | 32 | `jacobian` | 76.29 us | 73.25 us | +4.2% |
| PCS | 32 | `jacobian` | 197.54 us | 191.58 us | +3.1% |
| GVS | 32 | `jacobian_inertialframe` | 2068.54 us | 2019.71 us | +2.4% |
| PlanarPCS | 32 | `total_energy` | 1146.92 us | 1116.38 us | +2.7% |
| PCS | 32 | `total_energy` | 3867.50 us | 3581.17 us | +8.0% |
| GVS | 32 | `total_energy` | 24889.42 us | 24719.58 us | +0.7% |

Except for the PlanarPCS tip `forward_kinematics` microcase, the forward-only
custom-JVP wrapper overhead is generally around 0-5% under JIT, with PCS
`total_energy` at about 8%. The `total_energy` protected forward pass was also
fixed so `_total_energy` calls protected child hooks directly rather than
re-entering public custom-JVP wrappers through `kinetic_energy` and
`potential_energy`.

## System Methods vs `origin/main`

Speedup is `origin/main execution time / current execution time`, so values above
`1.00x` mean this branch was faster and values below `1.00x` mean this branch was
slower. These timings use the default behavior on this branch, where custom JVPs
are enabled.

| system | geomean current/main | n=32 geomean | strongest method | weakest method |
|---|---:|---:|---|---|
| PlanarPCS | 1.07x | 0.97x | `jacobian` 1.18x | `total_energy` 0.93x |
| PCS | 1.05x | 0.97x | `jacobian` 1.24x | `jacobian_and_time_derivative` 0.92x |
| GVS | 1.01x | 1.01x | `jacobian_inertialframe` 1.14x | `gravitational_force` 0.92x |

Per-method execution geomean, again using `origin/main / current`:

| method | PlanarPCS | PCS | GVS |
|---|---:|---:|---:|
| `forward_kinematics` | 1.08x | 1.07x | 1.06x |
| `jacobian` / frame-specific Jacobian | 1.18x | 1.24x | 0.98-1.14x |
| `jacobian_and_time_derivative` | 1.04x | 0.92x | 0.94-1.09x |
| `forward_dynamics` | 1.14x | 1.06x | 0.98x |
| `total_energy` | 0.93x | 1.01x | 0.99x |

## Conclusion

The custom-JVP derivative paths improve several important derivative workloads,
especially arc-length Jacobian JVPs and kinetic/total energy gradients at larger
segment counts. The benefit is not uniform: a few small or simple gradient paths
remain neutral or slower, but the aggregate derivative benchmark is faster with
custom JVPs enabled for PlanarPCS, PCS, and GVS.

After fixing `benchmark_system_methods.py` to actually JIT each measured method,
the previous apparent large forward-pass regression disappears. Standard system
methods are roughly at parity with `origin/main`: geomean execution speedups are
1.07x for PlanarPCS, 1.05x for PCS, and 1.01x for GVS. The remaining differences
are small enough to interpret as method-specific tradeoffs and benchmark noise
rather than a broad forward-pass regression.

## Validation

- `uv run python -m py_compile src/soromox/systems/gvs/strain_bases.py`
- `uv run ruff check src/soromox/systems/gvs/strain_bases.py`
- `uv run python -m py_compile src/soromox/systems/soft_robot.py tools/benchmarks/benchmark_system_methods.py`
- `uv run ruff check src/soromox/systems/soft_robot.py tools/benchmarks/benchmark_system_methods.py`
- `git diff --check -- src/soromox/systems/soft_robot.py tools/benchmarks/benchmark_system_methods.py`
- `uv run pytest tests/systems/test_soft_robot_defaults.py -q`
- Focused GVS derivative regression tests:

```bash
uv run pytest \
  tests/systems/test_gvs.py::test_gvs_arc_length_derivatives_match_autodiff \
  tests/systems/test_gvs.py::test_gvs_custom_jvps_include_arc_length_derivative \
  tests/systems/test_gvs.py::test_strain_basis_consistency_kinematics_jacobians_and_dynamics \
  -q
```

Benchmark artifacts:

- `benchmarks/final-derivative-custom-jvp.csv`
- `benchmarks/final-derivative-custom-jvp.json`
- `benchmarks/final-derivative-custom-jvp.md`
- `benchmarks/final-system-methods-current.csv`
- `benchmarks/final-system-methods-current.json`
- `benchmarks/final-system-methods-current.pdf`
- `benchmarks/final-system-methods-main.csv`
- `benchmarks/final-system-methods-main.json`
- `benchmarks/final-system-methods-main.pdf`

### CODE REVIEW

**Review by mohammedtarnini — APPROVED** [review]
Changes look good where we can use this framework to perform our custom analytical derivatives for some components. I understood that in soft_robot.py, there is a check whether custom_jvp is enabled or not and based on that it will run functions that compute custom derivatives for forward kinematics, jacobian, kinetic energy, elastic energy, and total energy. However, I believe that the main contribution for us is the derivative of forward dynamics not the aforementioned components, which is not currently present.
